### ISSUE TITLE
`:q` the configuration with an open tab crashes oni

### ISSUE THREAD

**Issue by tcoopman** [onivim/oni2#300]
Steps to reproduce:

1. Open oni
2. Type something in the current tab
3. Open configuration
4. Press `:q`

Result: oni hangs and then crashes.

Output in terminal:

```bash
[DEBUG] Message length: 72
Scope loaded: source.json
[DEBUG] Message length: 711
Request timed out: nvim_buf_attach (30)[1, true, {}]
Queued responses: 
Fatal error: exception Oni_Neovim__NeovimApi.RequestFailed("Request timed out: nvim_buf_attach (30)[1, true, {}]")
```

**Comment by kristopher-h** [comment]
I poked around a bit with regards to this issue and added some extra debug prints to see what is going on. It seems to happen when there is unsaved changes in a buffer that is currently not the open one. The flow seems to be something like (debug print included below):

1. :q issued
2. BufUnload autocmd (active buffer) 
3. BufEnter autocmd (changing to the buffer with unsaved changes)
4. Redraw triggered ("No write since last change ... press ENTER or type command to continue")
5. nvim_buf_attach sent from oni (which causes the timeout due to no reply)

<pre>
[DEBUG] Got message: |[2, redraw,
 [[wildmenu_hide, []], [cmdline_show, [[[0, q]], 1, :, , 0, 1]], [flush, []]]]|
[DEBUG] Sent message: |[0, 17, nvim_input, [<CR>]]|
[DEBUG] Got message: |[1, 17, (), 4]|
[DEBUG] Got message: |[2, redraw, [[wildmenu_hide, []], [flush, []]]]|
[DEBUG] Got message: |[2, redraw,
 [[cmdline_hide, [1]],
  [grid_line,
   [1, 19, 0,
    [[E, 4], [3], [7], [:], [ ], [N], [o], [ ], [w], [r], [i], [t], [
     e], [ ], [s], [i], [n], [c], [e]]], [1, 18, 0, [[ , 9, 20]]]],
  [grid_scroll, [1, 17, 20, 0, 20, 1, 0]],
  [grid_line, [1, 19, 0, [[ , 0, 20]]], [1, 18, 19, [[ , 4]]],
   [1, 19, 0, [[l, 4], [a], [s], [t], [ ], [c], [h], [a], [n], [g], [e]]]],
  [grid_cursor_goto, [1, 19, 11]], [mode_change, [normal, 0]], [flush, []]]]|
[DEBUG] Got message: |[2, oni_plugin_notify, [[autocmd, BufUnload, [2, 1, 1, 0]]]]|
[DEBUG] Got message: |[2, nvim_buf_detach_event, [(0 "\002")]]|
[DEBUG] Got message: |[2, oni_plugin_notify, [[autocmd, BufEnter, [1, 1, 1, 1]]]]|
[DEBUG] Got message: |[2, oni_plugin_notify, [[autocmd, BufWinEnter, [1, 1, 1, 1]]]]|
[DEBUG] Got message: |[2, redraw,
 [[grid_scroll, [1, 16, 20, 0, 20, 1, 0]],
  [grid_line, [1, 19, 0, [[ , 0, 20]]],
   [1, 19, 0,
    [[E, 4], [1], [6], [2], [:], [ ], [N], [o], [ ], [w], [r], [i], [
     t], [e], [ ], [s], [i], [n], [c]]]],
  [grid_scroll, [1, 15, 20, 0, 20, 1, 0]],
  [grid_line, [1, 19, 0, [[ , 0, 20]]], [1, 18, 19, [[e, 4]]],
   [1, 19, 0,
    [[ , 4], [l], [a], [s], [t], [ ], [c], [h], [a], [n], [g], [e], [
      ], [f], [o], [r], [ ], [b], [u]]]],
  [grid_scroll, [1, 14, 20, 0, 20, 1, 0]],
  [grid_line, [1, 19, 0, [[ , 0, 20]]], [1, 18, 19, [[f, 4]]],
   [1, 19, 0,
    [[f, 4], [e], [r], [ ], ["], [[], [N], [o], [ ], [N], [a], [m], [
     e], []], ["]]]], [grid_scroll, [1, 13, 20, 0, 20, 1, 0]],
  [grid_line, [1, 19, 0, [[ , 0, 20]]],
   [1, 19, 0,
    [[P, 27], [r], [e], [s, 27, 2], [ ], [E], [N], [T], [E], [R], [ ], [
     o], [r], [ ], [t], [y], [p], [e]]]],
  [grid_scroll, [1, 12, 20, 0, 20, 1, 0]],
  [grid_line, [1, 19, 0, [[ , 0, 20]]], [1, 18, 19, [[ , 27]]],
   [1, 19, 0,
    [[c, 27], [o], [m, 27, 2], [a], [n], [d], [ ], [t], [o], [ ], [c], [
     o], [n], [t], [i], [n], [u], [e]]]], [grid_cursor_goto, [1, 19, 19]],
  [flush, []]]]|
[DEBUG] Sent message: |[0, 18, nvim_buf_attach, [1, true, {}]]|
Request timed out: nvim_buf_attach (18)[1, true, {}]
Queued responses: 
Fatal error: exception Oni_Neovim__NeovimApi.RequestFailed("Request timed out: nvim_buf_attach (18)[1, true, {}]")
Raised at file "src/editor/Neovim/NeovimApi.re", line 178, characters 8-43
Called from file "src/editor/Neovim/NeovimProtocol.re", line 73, characters 6-115
Called from file "src/editor/Store/NeovimStoreConnector.re", line 217, characters 12-52
</pre>

A similar issue happens if ":q" is run when there are unsaved changes in the current window. But since there is no "BufEnter" triggered nothing happens until a new request is sent. Example to trigger this:

1. open oni
2. Make a change in current buffer and press ESC
3. Run ":q"
4. Press ctrl-p and open README.md

I have no clue how this would/should be fixed but maybe the added information can be of help. 

I added the following debug prints:
<pre>
diff --git a/src/editor/Neovim/NeovimApi.re b/src/editor/Neovim/NeovimApi.re
index f32d5d6..266b571 100644
--- a/src/editor/Neovim/NeovimApi.re
+++ b/src/editor/Neovim/NeovimApi.re
@@ -92,4 +92,4 @@ let make = (msgpack: MsgpackTransport.t) => {
    */
-  let handleMessage = (m: M.t) =>
-    /* prerr_endline ("Got message: |" ++ M.show(m) ++ "|"); */
+  let handleMessage = (m: M.t) => {
+    Log.debug("Got message: |" ++ M.show(m) ++ "|");
     switch (m) {
@@ -111,2 +111,3 @@ let make = (msgpack: MsgpackTransport.t) => {
     };
+  }
 
@@ -134,2 +135,3 @@ let make = (msgpack: MsgpackTransport.t) => {
     msgpack.write(request);
+    Log.debug("Sent message: |" ++ M.show(request) ++ "|");
   };
@@ -145,2 +147,3 @@ let make = (msgpack: MsgpackTransport.t) => {
       msgpack.write(request);
+      Log.debug("Sent message: |" ++ M.show(request) ++ "|");
</pre>

**Comment by bryphe** [comment]
This should be fixed by #391 now 👍
